lib/internal/loader/ModuleJob.js

```diff
@@ -1,91 +1,93 @@
         return (await dependencyJob.modulePromise).module;
       });
+      if (enableDebug) {
+        // Make sure all dependencies are entered into the list synchronously.
+        Object.freeze(dependencyJobs);
+      }
       return SafePromise.all(dependencyJobs);
     };
-    this.linked = linked();
+    // Promise for the list of all dependencyJobs.
+    this.linked = link();
 
     // instantiated == deep dependency jobs wrappers instantiated,
     // module wrapper instantiated
     this.instantiated = undefined;
   }
 
-  instantiate() {
+  async instantiate() {
     if (this.instantiated) {
       return this.instantiated;
     }
-    return this.instantiated = new Promise(async (resolve, reject) => {
-      const jobsInGraph = new SafeSet();
-      let jobsReadyToInstantiate = 0;
-      // (this must be sync for counter to work)
-      const queueJob = (moduleJob) => {
-        if (jobsInGraph.has(moduleJob)) {
-          return;
-        }
-        jobsInGraph.add(moduleJob);
-        moduleJob.linked.then((dependencyJobs) => {
-          for (const dependencyJob of dependencyJobs) {
-            queueJob(dependencyJob);
-          }
-          checkComplete();
-        }, (e) => {
-          if (!this.hadError) {
-            this.error = e;
-            this.hadError = true;
-          }
-          checkComplete();
-        });
-      };
-      const checkComplete = () => {
-        if (++jobsReadyToInstantiate === jobsInGraph.size) {
-          // I believe we only throw once the whole tree is finished loading?
-          // or should the error bail early, leaving entire tree to still load?
-          if (this.hadError) {
-            reject(this.error);
-          } else {
-            try {
-              this.module.instantiate();
-              for (const dependencyJob of jobsInGraph) {
-                dependencyJob.instantiated = resolvedPromise;
-              }
-              resolve(this.module);
-            } catch (e) {
-              e.stack;
-              reject(e);
-            }
-          }
-        }
-      };
-      queueJob(this);
-    });
+    return this.instantiated = this._instantiate();
+  }
+
+  // This method instantiates the module associated with this job and its
+  // entire dependency graph, i.e. creates all the module namespaces and the
+  // exported/imported variables.
+  async _instantiate() {
+    const jobsInGraph = new SafeSet();
+
+    const addJobsToDependencyGraph = async (moduleJob) => {
+      if (jobsInGraph.has(moduleJob)) {
+        return;
+      }
+      jobsInGraph.add(moduleJob);
+      const dependencyJobs = await moduleJob.linked;
+      return Promise.all(dependencyJobs.map(addJobsToDependencyGraph));
+    };
+    try {
+      await addJobsToDependencyGraph(this);
+    } catch (e) {
+      if (!this.hadError) {
+        this.error = e;
+        this.hadError = true;
+      }
+      throw e;
+    }
+    this.module.instantiate();
```

Comment:
Amazing to see this simplification. Not having to call instantiate for each module in the right order is a much nicer v8 API!

Reply:
Just in case it wasn’t clear, we only had one call to `instantiate()` before as well, but maybe this is a bit more obvious now :)

Reply:
Huh, it seems I was simply trying to post-justify my less elegant approach :)

Reply:
Heh, didn’t know this was your code ;)